build: enable missing OpenSSF-recommended warnings, with fixes

vszakats opened the pull request:
https://best.openssf.org/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C++.html
[as of [2023-11-29](https://github.com/ossf/wg-best-practices-os-developers/blob/56c0fde3895bfc55c8a973ef49a2572c507b2ae1/docs/Compiler-Hardening-Guides/Compiler-Options-Hardening-Guide-for-C-and-C%2B%2B.md)]

Enable new recommended warnings (except `-Wsign-conversion`):

- enable `-Wformat=2` for clang (in both cmake and autotools).
- add `CURL_PRINTF()` internal attribute and mark functions accepting
  printf arguments with it. This is a copy of existing
  `CURL_TEMP_PRINTF()` but using `__printf__` to make it compatible
  with redefinting the `printf` symbol:
  https://gcc.gnu.org/onlinedocs/gcc-3.0.4/gcc_5.html#SEC94
- fix `CURL_PRINTF()` and existing `CURL_TEMP_PRINTF()` for
  mingw-w64 and enable it on this platform.
- enable `-Wimplicit-fallthrough`.
- enable `-Wtrampolines`.
- add `-Wsign-conversion` commented with a FIXME.
- cmake: enable `-pedantic-errors` the way we do it with autotools.
  Follow-up to d5c0351055d5709da8f3e16c91348092fdb481aa #2747
- lib/curl_trc.h: use `CURL_FORMAT()`, this also fixes it to enable format
  checks. Previously it was always disabled due to the internal `printf`
  macro.

Fix them:

- fix bug where an `set_ipv6_v6only()` call was missed in builds with
  `--disable-verbose` / `CURL_DISABLE_VERBOSE_STRINGS=ON`.
- add internal `FALLTHROUGH()` macro.
- replace obsolete fall-through comments with `FALLTHROUGH()`.
- fix fallthrough markups: Delete redundant ones (showing up as
  warnings in most cases). Add missing ones. Fix indentation.
- silence `-Wformat-nonliteral` warnings with llvm/clang.
- fix one `-Wformat-nonliteral` warning.
- fix new `-Wformat` and `-Wformat-security` warnings.
- fix `CURL_FORMAT_SOCKET_T` value for mingw-w64. Also move its
  definition to `lib/curl_setup.h` allowing use in `tests/server`.
- lib: fix two wrongly passed string arguments in log outputs.
  Co-authored-by: Jay Satiro
- fix new `-Wformat` warnings on mingw-w64.

Closes #12489

---

TODO:
- [x] fix to enable format checks for mingw.
- [x] fix `-Warith-conversion` warnings. [other PR]
- [x] fix `-Wsign-conversion` warnings. [other PR]

bagder commented on include/curl/curl.h:
```diff
@@ -133,6 +133,15 @@ typedef void CURLSH;
+#else
+#  define CURL_FALLTHROUGH()  do {} while (0)
+#endif
+#endif
```
Ah right, is just there to make this macro use work...

vszakats replied:
Good point about the placement, I didn't gave a thought at all (and that while copying this from libssh2's internal header). Definitely needs to be local, I'll move it.

vszakats replied:
examples are also compiled in CI: https://github.com/curl/curl/actions/runs/7146212736/job/19463469167?pr=12489#step:32:36

bagder replied:
Yes, but I don't think that is reason enough for us to add this to the public headers. We should probably rather try to
1. avoid the fallthroughs in example code or
2. avoid the warning in example code

bagder replied:
I could add that I previously have avoided this particular warning just because I think this "clutter" is bad for readability.

vszakats replied:
The clutter was already there as `/* FALLTHROUGH */` comments, the problem was that newer compiler versions no longer honored it, so we had to disable the warning for them. This has been solved since on the compiler end.

vszakats replied:
Looking into the code, it was easy to avoid those fallthroughs.

bagder replied:
I think /* FALLTHROUGH */ comments was a rather elegant way. Oh well, adapt and overcome.

vszakats replied:
Doc [lists](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wimplicit-fallthrough_003d) 11 lines of regexps to define the accepted legacy syntax, spread out to 4 different modes. And this is just gcc. Just grepping them was non-trivial! I'm happy gcc and llvm could agree on a single, common syntax.

bagder commented on lib/cf-socket.c:
```diff
@@ -983,20 +983,14 @@ static CURLcode cf_socket_open(struct Curl_cfilter *cf,
-    const char *ipmsg;
 #ifdef ENABLE_IPV6
-    if(ctx->addr.family == AF_INET6) {
-      set_ipv6_v6only(ctx->sock, 0);
```
I don't think it is

vszakats replied:
Thanks, the fix is already there, and I will document it in the commit message.

vszakats commented:
This is ready now.

bagder commented on lib/cf-h1-proxy.c:
```diff
@@ -183,7 +183,7 @@ static void h1_tunnel_go_state(struct Curl_cfilter *cf,
     data->state.authproxy.done = TRUE;
     data->state.authproxy.multipass = FALSE;
-    /* FALLTHROUGH */
+    CURL_FALLTHROUGH();
```
I think this could drop the `CURL_` prefix since this is an internal macro. To keep it shorter and more similar to what it used to be.

vszakats replied:
My preference goes for a prefix because it says cleary where the symbol is coming from, and also avoids the minor chance of colliding with a symbol coming from a 3rd-party header. We also already use the CURL prefix (or at least a fairly unique name) for most internal symbols. This one is used 85 times in the codebase, so relatively rare.

Are we okay with these?

vszakats commented:
Renamed to use `FALLTHROUGH();`.

Is there something else to address?

vszakats commented:
Disabling these two errors by not passing the non-char* strings for these error log outputs:
```
lib/socks_sspi.c:334:15: warning: format '%s' expects argument of type 'char *', but argument 3 has type 'SEC_WCHAR *' {aka 'short unsigned int *'} [-Wformat=]
lib/vtls/schannel.c:693:63: warning: format '%s' expects argument of type 'char *', but argument 4 has type 'TCHAR *' {aka 'short unsigned int *'} [-Wformat=]
```

They will need subsequent fixes to re-add the strings to them.

jay commented:
I'm not a fan of FIXMEs in the code like that they are easily forgotten. I will see about converting them

vszakats commented:
Thank you @jay!

vszakats commented:
The printf format checking "journey":

My goal was to enable it for internal functions and with the mingw-w64 compiler. This presented some issues, here's a summary:

1. The existing `CURL_TEMP_PRINTF()` macro (in `curl/mprintf.h`) gets undefined after use, so had to add a new, persistent one to the internal headers.
2. A one to one copy did not work because curl's headers redefine the symbol `printf` which happens to be the keyword we need to pass to `__attribute__(format())`. The solution is to pass the alternate keyword: `__printf__`. This has been around for ages, though probably a little younger than `printf`.
3. I kept `printf` in the public (temporary) macro, though IMO it'd be cleaner to use the underscored flavour there too. The compatibility cost seems negligible, but would need double checking in the GNU C source code to confirm.
4. This revealed the first round of legitimate warnings.
5. The format attribute has been historically disabled for `__MINGW32__` in curl, thus the next task was to enable it.
6. In the first test, as expected, mingw started spitting warnings all over the place with `__printf__`.
7. I tried `__MINGW_PRINTF_FORMAT` instead.
8. This needs a fallback because it was introduced in mingw-w64 3.0.
9. CI results were already useful to fix Windows-specific format issues.
10. At this point mingw (tested with gcc) did not like the `%qd` format specifier in `CURL_FORMAT_SOCKET_T`. Fixed that by replacing it with `%zd`.
11. Then it turned out while testing locally, that it doesn't work with `__clang__` because for this compiler, the mingw-w64 header maps the `__MINGW_PRINTF_FORMAT` macro to `printf` which conflicts with the local macro, as seen earlier. The solution was to map it manually to `__printf__` for `__clang__`. [ci](https://ci.appveyor.com/project/curlorg/curl/builds/48764052)
12. The next problem was that depending on GNU C version and/or mingw-w64 version, `__MINGW_PRINTF_FORMAT` gets mapped to `gnu_printf` in newer releases (gcc v13 job in AppVeyor), which works without warnings, but for gcc v6, v7, v9 it maps to `ms_printf`, which triggers warnings for `%z`. I fixed this by abandoning `__MINGW_PRINTF_FORMAT` and using `gnu_printf` straight, for mingw + gcc.
13. This triggered another problem where curl sources use MS-specific formats for `size_t` types, raising warnings with old mingw + gcc (v6, v7, v9 again) releases. [ci](https://ci.appveyor.com/project/curlorg/curl/builds/48764281) I fixed this by redefining `CURL_FORMAT_CURL_OFF_T[U]` for mingw to the GNU format. (Another option would be to disable format checks for older GCC versions.)

48 format strings had issues after extending verification to new functions.
37 format additional strings had issues when enabled for mingw (19 of them in libcurl, the rest in tests).

In summary: The idea is that libcurl, the curl tool and tests are all using the local `printf()` implementation (from `lib/mprintf.c`), which speaks the GNU format (and understands some MS extensions). The formats used in code must comply with that (and not whatever the CRT's `printf()` speaks). Then setup the format checker compiler attribute for the GNU format, where this isn't the default or needs a special keyword (`gnu_format`).

I hope this makes sense or maybe even useful. Any remarks or comments are welcome, especially if it points to any flaw in the above, before merging this.

bagder commented:
:balloon: very nice work on this @vszakats!
